api/src/main/resources/marquez/db/migration/V41__alter_jobs_add_job_parent_id.sql

```diff
@@ -0,0 +1,33 @@
+ALTER TABLE jobs ADD COLUMN IF NOT EXISTS parent_job_id uuid CONSTRAINT jobs_parent_fk_jobs REFERENCES jobs (uuid);
+ALTER TABLE runs ADD COLUMN IF NOT EXISTS parent_run_id uuid CONSTRAINT runs_parent_fk_runs REFERENCES runs (uuid);
+DROP INDEX IF EXISTS jobs_name_index;
+CREATE UNIQUE INDEX IF NOT EXISTS jobs_name_parent ON jobs (name, namespace_name, parent_job_id) INCLUDE (uuid, name, namespace_name, parent_job_id);
+
+CREATE OR REPLACE VIEW jobs_view
+AS
+    with recursive job_fqn AS (
+        SELECT uuid, name, namespace_name, parent_job_id
+        FROM jobs
+        UNION ALL
+        SELECT j1.uuid,
+               CASE WHEN j2.name IS NOT NULL THEN j2.name || '.' || j1.name ELSE j1.name END AS name,
+               CASE WHEN j2.namespace_name IS NOT NULL THEN j2.namespace_name ELSE j1.namespace_name END AS namespace_name,
+               j2.parent_job_id
+        FROM jobs j1
+        INNER JOIN job_fqn j2 ON j2.uuid=j1.parent_job_id
+    )
+    SELECT f.uuid,
+           f.name,
+           f.namespace_name,
+           j.name AS simple_name,
+           j.parent_job_id,
+           j.type,
+           j.created_at,
+           j.updated_at,
+           j.namespace_uuid,
+           j.description,
+           j.current_version_uuid,
+           j.current_job_context_uuid,
+           j.current_location,
+           j.current_inputs
+           FROM job_fqn f, jobs j WHERE j.uuid=f.uuid
```

Comment:
What are your thoughts on being more explicit with the job _fully qualified name_? That is, if someone was querying the `jobs` table, the `name` column would be the _simple_ name, but when querying `jobs_view` the name is now the fully qualified name (though the `name` column is renamed for added context).

I find it confusing to repurpose the `name` column and would prefer to add a `fully_qualified_name` to `job_views`. That said, that comes with major drawbacks given the renaming nightmare. If we consider the response:

```
{
  "id": {
  "namespace": "my-namespace",
  "name": "parent-job.my-job"
  },
  "type": "BATCH",
  "name": "parent-job.my-job",
  .
  .
```

`name` in this context would be the fully qualified name, and as outlined in your [proposal](https://github.com/MarquezProject/marquez/issues/1928) we'd add a `simpleName` field. I find this reasonable.   This change would also mean, `jobId`s will now use the fully qualified name. So maybe I talked myself out of using `fully_qualified_name` 😅 

I guess my thoughts are that the view does provide guarantees for the migration, but we are left with long term "maintenance". For example, any columns added to `jobs` would also be added to `jobs_view`, any newcomers to the project would have to be aware of this view, etc. I think this is why initially I preferred `option 2` in your propsoal:

> 2. Do not backfill the jobs or runs tables. Jobs and runs can continue to use the simple names in tables and those names will be translated at runtime when the API response is constructed.

Reply:
That said, we don't expect the db model to be queried directly and the API model is the public contract. I'm just wondering if there's a scenario where a `simple_name` column is added to `jobs` and we backfill the `name` column in place if the `parent_job_uuid` is not `null`. If `null` both `name` and `simple_name` would contain the same value.

Reply:
But, you've also thought about this in more detail than I have! The goal is to support job grouping, so if that means we add views and ensure we minimize user downtime, then no objections on my end. Just thought I'd bring up a few things but certainly not a blocker to move forward.

Reply:
Option 2 in the proposal requires either querying the view or changing the code to query the correct tables/columns. E.g., if a request comes in for `parent-job.my-job`, we can't simply query the unaltered `jobs` table, since the `name` column will only contain `my-job`. So the either the code will need to parse the job name (very bad idea, given that there are jobs whose names already contain a dot character) or it will have to query a table/view that has the FQN pre-constructed. So even if we did absolutely no backfilling, the view is likely the _most_ maintainable option going forward. To get away with no view at all requires option 1 - update all the job name in the `jobs`, `job_version`, `runs`, and `lineage_events` tables. We can add the `simple_name` column for the `jobs` table, but everything else would be the FQN. Assuming that migration is successful, there's 0 maintenance going forward. But that migration is potentially expensive.